File: src/main/java/com/dingdangmaoup/resin/pura/resin/ResinPersistentDataHelper.java
package com.dingdangmaoup.resin.pura.resin;

import com.dingdangmaoup.resin.pura.ResinPersistentData;
import com.dingdangmaoup.resin.pura.resin.configuration.JmxConfigurationStrategy;
import com.dingdangmaoup.resin.pura.resin.configuration.ResinConfigurationStrategy;
import com.intellij.execution.ExecutionException;
import com.intellij.javaee.appServers.appServerIntegrations.ApplicationServer;
import com.intellij.openapi.diagnostic.Logger;
import org.jetbrains.annotations.Nullable;

/**
 * @author michael.golubev
 */
public class ResinPersistentDataHelper {

  private static final Logger LOG = Logger.getInstance(ResinPersistentDataHelper.class);

  private final ApplicationServer myApplicationServer;

  public ResinPersistentDataHelper(ApplicationServer applicationServer) {
    myApplicationServer = applicationServer;
  }

  @Nullable
  public ResinPersistentData getPersistentData() {
    return myApplicationServer != null ? (ResinPersistentData)myApplicationServer.getPersistentData() : null;
  }

  @Nullable
  public ResinInstallation getInstallation() {
    try {
      return getInstallationOrError();
    }
    catch (ExecutionException e) {
      LOG.debug(e);
      return null;
    }
  }

  @Nullable
  public ResinInstallation getInstallationOrError() throws ExecutionException {
    ResinPersistentData persistentData = getPersistentData();
    return persistentData != null ? ResinInstallation.create(persistentData.RESIN_HOME) : null;
  }

  @Nullable
  ResinConfigurationStrategy getStrategy() {
    ResinInstallation installation = getInstallation();
    return installation == null ? null : ResinConfigurationStrategy.getForInstallation(installation);
  }

  @Nullable
  public JmxConfigurationStrategy getJmxStrategy() {
    ResinConfigurationStrategy strategy = getStrategy();
    return strategy instanceof JmxConfigurationStrategy ? (JmxConfigurationStrategy)strategy : null;
  }

  public boolean hasJmxStrategy() {
    return getJmxStrategy() != null;
  }
}

File: src/main/kotlin/com/dingdangmaoup/resin/pura/resin/ResinPersistentDataHelper.kt
package com.dingdangmaoup.resin.pura.resin

import com.dingdangmaoup.resin.pura.ResinPersistentData
import com.dingdangmaoup.resin.pura.resin.configuration.JmxConfigurationStrategy
import com.dingdangmaoup.resin.pura.resin.jmx.JmxDeploymentClient
import com.intellij.execution.ExecutionException
import com.intellij.javaee.appServers.appServerIntegrations.ApplicationServer
import com.intellij.openapi.diagnostic.Logger
import java.io.File
import java.nio.file.Files
import java.nio.file.attribute.BasicFileAttributes
import java.io.IOException

class ResinPersistentDataHelper(private val applicationServer: ApplicationServer?) {
    private var myInstallationCache: InstallationCache? = null

    fun getPersistentData(): ResinPersistentData? {
        return applicationServer?.persistentData as? ResinPersistentData
    }

    fun getInstallation(): ResinInstallation? {
        return try {
            getInstallationOrError()
        } catch (e: ExecutionException) {
            LOG.debug(e)
            null
        }
    }

    @Throws(ExecutionException::class)
    @Synchronized
    fun getInstallationOrError(): ResinInstallation? {
        val persistentData = getPersistentData()
        if (persistentData == null) {
            myInstallationCache = null
            return null
        }

        val homePath = persistentData.RESIN_HOME
        val cached = myInstallationCache
        if (cached != null && cached.homePath == homePath && cached.hasRequiredDirectories() &&
            cached.fingerprint == installationFingerprint(File(homePath))) {
            return cached.installation
        }

        // Cache only a validated installation. An invalid path may become valid later
        // without changing the persisted value, so failed creations must remain retryable.
        myInstallationCache = null
        val installation = ResinInstallation.create(homePath)
        myInstallationCache = InstallationCache(homePath, installation, installationFingerprint(File(homePath)))
        return installation
    }

    fun getJmxStrategy(): JmxConfigurationStrategy? {
        val installation = getInstallation() ?: return null
        if (!installation.getVersion().allowJmx()) return null
        return JmxDeploymentClient(installation)
    }

    fun hasJmxStrategy(): Boolean {
        return getJmxStrategy() != null
    }

    companion object {
        private val LOG = Logger.getInstance(ResinPersistentDataHelper::class.java)

        private fun installationFingerprint(home: File): List<String?> =
            listOf("resin.jar", "jsdk23.jar", "jsdk-24.jar").map { name ->
                try {
                    val attributes = Files.readAttributes(File(home, "lib/$name").toPath(), BasicFileAttributes::class.java)
                    "${attributes.fileKey()}:${attributes.size()}:${attributes.lastModifiedTime()}"
                } catch (_: IOException) {
                    null
                }
            }
    }

    private data class InstallationCache(
        val homePath: String,
        val installation: ResinInstallation,
        val fingerprint: List<String?>,
    ) {
        fun hasRequiredDirectories(): Boolean {
            val home = installation.getResinHome()
            return home.isDirectory && File(home, "bin").isDirectory && File(home, "lib").isDirectory
        }
    }
}
